docs/modules/ROOT/pages/references/architecture/control-api-invitation.adoc

```diff
@@ -0,0 +1,82 @@
+= {controlapi}: Invitation
+
+TIP: This resource implements the xref:references/functional-requirements/portal.adoc#_feature_send_invitations["Send Invitations"] and xref:references/functional-requirements/portal.adoc#_feature_accept_invitations["Accept Invitations"] features.
+
+== Object
+
+A virtual resource is chosen to allow for RBAC filtering for list operations.
+The `Invitation` is stored in a `Secret` resource since it contains sensitive information such as the invitation token.
+
+.Virtual resource
+[source,yaml]
+----
+apiVersion: users.appuio.io/v1
+kind: Invitation
+metadata:
+  name: E303B166-5D66-4151-8F5F-B84BA84A7559 <1>
+spec:
+  note: "New employee dev1 (Delilah Vernon) starting 2020-04-01" <2>
+  email: "dev1@acme.com" <3>
+  # For billing entity invitations
+  roleBindingRef:
+    apiGroup: "rbac.authorization.k8s.io"
+    kind: "ClusterRoleBinding"
+    name: "be-123-viewer"
+    namespace: "" <4>
+  # OR
+  # For organization invitations
+  roleBindingRef:
+    apiGroup: "rbac.authorization.k8s.io"
+    kind: "RoleBinding"
+    name: "control-api:organization-admin"
+    namespace: "acme" <5>
+status:
+  token: "BL30rELat6Faee5OgfmmZGhnECnp5H8GCbw7PcgRlTTXW0OAJMYl5anzuIdO" <6>
+  conditions:
+  - lastTransitionTime: "2020-03-31T14:00:00Z"
+    status: "True"
+    type: EmailSent
+  - lastTransitionTime: "2020-04-01T14:00:00Z"
+    status: "True"
+    message: "Redeemed by acme-dev1"
+    type: Redeemed
+----
+<1> The unique identifier of the invitation.
+This is a random UUID.
+<2> Free-form text that can be used to add a note for the owner sending the invitation.
+<3> The email address of the user to invitation.
+<4> Billing entity RBAC is cluster-scoped with `resourceNames` set to the billing entity ID.
+<5> Organization RBAC is a namespaced `RoleBinding`.
+<6> The token that is used to accept the invitation.
+Generated by the API server.
+
+== Access Control
+
+[TIP]
+We use the same access control mechanism for `Invitation` as for xref:references/architecture/control-api-org.adoc#_access_control[`Organization` resources].
+
+Invitations can only reference `RoleBinding` resources that the user has access to.
+This is enforced by the API server and mitigates privilege escalation risks.
+
+== Accepting an Invitation
+
+Invitations can be accepted by any user in possession of the invitation token.
+
+Invitations are accepted by sending an `REDEEM` request to the `Invitation` resource.
+
+[source,bash]
+----
+INVITATION=E303B166-5D66-4151-8F5F-B84BA84A7559 <1>
+TOKEN=BL30rELat6Faee5OgfmmZGhnECnp5H8GCbw7PcgRlTTXW0OAJMYl5anzuIdO
+
+curl -XREDEEM "$API/apis/users.appuio.io/v1/invitation/$INVITATION/$TOKEN"
```

Comment:
Wow! I never heard of that... Do you have any links to learn more about that? Sounds interesting.

Reply:
It's a custom method as allowed by HTTP 1.1 and later https://www.rfc-editor.org/rfc/rfc2616#section-9

Which can be implemented by our apiserver and allows for roles such as

```
- apiGroups:
  - "users.appuio.io"
  resources:
  - invitations
  verbs:
  - redeem
```

Reply:
Would this also work with `kubectl`, for instance `kubectl redeem $INVITATION token=$TOKEN`?